### DISCUSSION TITLE
How do Shopify/Liquid and shopify.dev differ?

### DISCUSSION THREAD

**Discussion by sebastienros** [Shopify/liquid#1881] [Q&A] [answered]
I am often using the code blocks in https://shopify.dev/docs/api/liquid#liquid_basics to test some snippets of liquid. However it seems like the results vary with this repository. Here is an example, where `last` and `first` can be used on a string, but I can't find this logic in this repository.

![image](https://github.com/user-attachments/assets/a1648df9-fb52-4e40-a5e5-13c423f92c66)

Another example is identifiers, where on the shopify.dev you can use `$` as the first char, but in this repository the Regex for identifiers doesn't include it.

![image](https://github.com/user-attachments/assets/04d62b10-a1a5-4b1c-bae2-def5ffbfc1f1)

What are the reasons for these differences? Which one is correct?

**Comment by jg-rp** [comment, marked as answer]
As far as `first` and `last` go, they are handled here:

https://github.com/Shopify/liquid/blob/323951b36fb6a4f1b7e4394cd5746cd78b49791b/lib/liquid/variable_lookup.rb#L70

Best guess (hopefully someone at Shopify will correct me if I’m wrong), strings on shopify.dev are [active support strings](https://guides.rubyonrails.org/active_support_core_extensions.html), which do respond to `first` and `last`. Where as standard [Ruby strings](https://docs.ruby-lang.org/en/master/String.html) don’t.

So it’s the string objects that are different, not the Liquid engine. 🤷
